### ISSUE TITLE
LDClient::close() does not clean up the internal `instances` map.

### ISSUE THREAD

**Issue by valeriyo** [launchdarkly/android-client-sdk#108]
This sequence:
- init(...)
- close()
- init(...)

does not work. The second `init()` call will succeed, but the LD library will be non-operational.

This is happening because, LDClient has a static `instances` map, which isn't cleared by `close()`:
```
    private static Map<String, LDClient> instances = null;

    ...

    public void close() throws IOException {
        LDClient.closeInstances();
    }

    ...

    private static void closeInstances() {
        for (LDClient client : instances.values()) {
            client.closeInternal();
        }
    }
```

So, the second `init()` will find `instances` and will bail out early:

```
        if (instances != null) {
            Timber.w("LDClient.init() was called more than once! returning primary instance.");
            return new LDSuccessFuture<>(instances.get(LDConfig.primaryEnvironmentName));
        }
```

Please do `instances = null` at the end of `closeInstances()` so that library can be re-initialized. Thanks.

**Comment by valeriyo** [comment]
Currently, I have:

```
        ldClient.close()

        // the workaround:
        ldClient.javaClass
            .declaredFields
            .firstOrNull { it.name == "instances" }
            ?.apply {
              isAccessible = true
              set(null, null)
            }
```

it works, but I would rather not have this hack in my code. Thanks.

**Comment by gwhelanLD** [comment]
Hi @valeriyo,

Could you let us know your use case for re-initializing the client? In general this isn't a feature we consider supported on our mobile SDKs, as the SDK lifecycle is considered tied to the application's lifecycle. 

Thanks,
@gwhelanLD

**Comment by valeriyo** [comment]
Our app has an "environment" switch for Production vs Staging (with our server), and those are mapped to two Launchdarkly environments - which map to two different keys. So, when the app switches to another environment, everything that is environment-specific needs to be cleaned up. What's unusual about that?

**Comment by gwhelanLD** [comment]
This specific case is actually addressed separately with the multi-environment feature. Take a look at the docs [here](https://docs.launchdarkly.com/sdk/client-side/android#multiple-environments). You can set either Production or Staging as the primary instance, and identify the other with `LDConfig.setSecondaryMobileKeys(Map<String, String>)`. Then the secondary environment can be retrieved with `LDClient.getForMobileKey(String)`.

Hope this helps!
@gwhelanLD

**Comment by valeriyo** [comment]
@gwhelanLD No, the multi-environment feature doesn't mitigate the fact that `close()` doesn't clean stuff up.

**Comment by DylanLange** [comment]
Having the same issue with the project I'm on. We need to be able to dynamically configure the LD config and reinitialise the client each time it updates so that the new configuration takes effect.

Is there any reason why `instances` can't be set to null @gwhelanLD? I'd prefer to not have to reflectively set it to null like @valeriyo has resorted to.

Thanks

**Comment by valeriyo** [comment]
Alright, this is brain-dead. Is it so hard to add a line to make this right?

```
    private static void closeInstances() {
        for (LDClient client : instances.values()) {
            client.closeInternal();
        }
        instances = null
    }
 ```
 
 I was considering LD for a new project, but I will look elsewhere because you suck. Bye.

**Comment by eli-darkly** [comment]
Fixed this in the 3.1.8 release.
